BrokerClient

danadesrosiers opened the pull request:
The Helper classes seem a bit scattered and have some code duplication.  Also, I don't love the name "Helper" since it doesn't clearly convey the responsibility of the class.  In an effort to simplify this area, I created a BrokerClient class that is a proxy for all calls to the BrokerService.  This would replace all Helper classes.  I moved the DiscoverMessageHandlers() method into an extension method since it didn't seem like it belonged in a "client".

The `SubscribeAsync()` methods now take the handler for the type being subscribed to and the BrokerClient stores the handlers and executes the correct one when `ProcessMessageAsync` is called.

I changed the naming pattern from "RegisterMessageType" to "Subscribe".  I think this naming is more clear and matches the Pub/Sub terminology better.  What do you think?

I also added endpoints `SubscribeAsync()` and `UnsubscribeAsync()` to IBrokerService instead of using the existing `RegisterSubscriberAsync()`, etc.  These methods take a `ReferenceWrapper` instead of a `ServiceReference` or `ActorReference`.  I thought it would nice to help simplify the BrokerService interface.  I'm curious your thoughts on this.

When we have all the details settled, I can mark things Obsolete and update the package version.  Or, maybe it's time for a major version bump.  There is quite a lot of Obsolete code, it would be nice to clean it up.

loekd commented on ServiceFabric.PubSubActors/Helpers/SubscriberExtensions.cs:
```diff
@@ -0,0 +1,50 @@
+
+    public static class SubscriberExtensions
+    {
+        public static Dictionary<Type, Func<object, Task>> DiscoverMessageHandlers(this ISubscriber service)
```
This breaks compatibility with the Service-bootstrappers, when removing the `SubscriberServiceHelper`

loekd replied:
I'd prefer sticking to `ISubscriberService` and `ISubscriberActor` instead of introducing another interface users need to add.

danadesrosiers replied:
Yeah, I'll have to update the bootstrappers.  I introduced `ISubscriber` thinking it would make it possible for users to set up handlers on another class rather than on the Service itself.  I'm not sure how useful that would really be, but that's what I was thinking.  I can take it out and use the existing interfaces only.

loekd commented:
Changes are looking good. 
If we're going to remove obsolete code as well, I propose we do it in this new version.
Let's see if we can also remove the `Interfaces` library.

danadesrosiers commented:
Sounds good.  I'll get to work on removing obsolete code.

danadesrosiers commented:
I removed a whole lot of obsolete code and the Interfaces library.  I broke it into several commits, so hopefully it's easier to follow all the changes.  What are your thoughts on updating framework version support?  Should we remove .NET Framework v4.5 and v4.6, and add v4.7?  I personally only need .NET Standard 2.0.

loekd commented:
Unrelated, I'd like to merge [this pr](https://github.com/loekd/ServiceFabric.PubSubActors/pull/59). It fixes a string hash issue in dotnet core.

loekd commented:
Supporting .Net fx 4.7 doesn't really help, except maybe for some singletons right? My philosophy is that when building shared libraries, you'll get the best compatibility if you support the lowest version of the fx.

danadesrosiers commented:
I made some changes to the BrokerClient to support generic versions of SubscribeAsync() and UnsubscribeAsync(), which is nicer for users who are subscribing manually.  Also removed "flush" from Unsubscribe since it isn't used and updated base classes to use remoting V2.

I'll work on adding some migration guidance to the README and update the version.  I was definitely expecting this to be v8.

danadesrosiers commented:
In testing, I found things weren't working, so don't merge yet.  I tracked the problem to the HashingHelper being null in ServiceReferenceWrapper when trying to create the queue.  Probably got messed up going through the serializer or something.

loekd commented:
I made a mistake in the field initializer. There's a PR to fix this. 
https://github.com/loekd/ServiceFabric.PubSubActors/pull/60

loekd commented:
I made two mistakes, one in the field initializer and one setting a default value in the service interface. This should now be fixed.

danadesrosiers commented:
I fixed some conflicts and lost the approval of the pull request, so I can't do it myself, but I'm ready for merge.

loekd commented:
This version is now released! Thanks for this major contribution @danadesrosiers !! I really appreciate all your efforts.
